test/functional/change_stream_spec_tests.js

```diff
@@ -0,0 +1,258 @@
+      if (!result.error) {
+        throw err;
+      }
+
+      assertEquality(err, result.error);
+    };
+  }
+
+  function makeTestAPM(test) {
+    const expectedEvents = test.expectations;
+
+    return function testAPM(ctx, events) {
+      expectedEvents
+        .map(e => e.command_started_event)
+        .map(normalizeAPMEvent)
+        .forEach((expected, idx) => {
+          if (!events[idx]) {
+            throw new Error(
+              `Expected there to be an APM event at index ${idx}, but there was none`
+            );
+          }
+          const actual = EJSONToJSON(events[idx]);
+          assertEquality(actual, expected);
+        });
+    };
+  }
+
+  function makeTestFnRunOperations(test) {
+    const target = test.target;
+    const operations = test.operations;
+    const success = test.result.success || [];
+
+    return function testFnRunOperations(ctx) {
+      const changeStreamPipeline = test.changeStreamPipeline;
+      const changeStreamOptions = test.changeStreamOptions;
+      ctx.changeStream = ctx[target].watch(changeStreamPipeline, changeStreamOptions);
+
+      const changeStreamPromise = readAndCloseChangeStream(ctx.changeStream, success.length);
+      const operationsPromise = runOperations(ctx.gc, operations);
+
+      return Promise.all([changeStreamPromise, operationsPromise]).then(args => args[0]);
+    };
+  }
+
+  function readAndCloseChangeStream(changeStream, numChanges) {
+    const close = makeChangeStreamCloseFn(changeStream);
+    let changeStreamPromise = changeStream.next().then(r => [r]);
+
+    for (let i = 1; i < numChanges; i += 1) {
+      changeStreamPromise = changeStreamPromise.then(results => {
+        return changeStream.next().then(result => {
+          results.push(result);
+          return results;
+        });
+      });
+    }
+
+    return changeStreamPromise.then(result => close(null, result), err => close(err));
+  }
+
+  function runOperations(client, operations) {
+    return operations
+      .map(op => makeOperation(client, op))
+      .reduce((p, op) => p.then(op), delay(200));
+  }
+
+  function makeChangeStreamCloseFn(changeStream) {
+    return function close(error, value) {
+      return new Promise((resolve, reject) => {
+        changeStream.close(err => {
+          if (error || err) {
+            return reject(error || err);
+          }
+          return resolve(value);
+        });
+      });
+    };
+  }
+
+  function normalizeAPMEvent(raw) {
+    return Object.keys(raw).reduce((agg, key) => {
+      agg[camelCase(key)] = raw[key];
+      return agg;
+    }, {});
+  }
+
+  function makeOperation(client, op) {
+    const target = client.db(op.database).collection(op.collection);
+    const command = op.name;
+    const args = [];
+    if (op.arguments && op.arguments.document) {
+      args.push(op.arguments.document);
+    }
+    return () => target[command].apply(target, args);
+  }
+
+  function assertEquality(actual, expected) {
```

Comment:
can we use/improve [this](https://github.com/mongodb/node-mongodb-native/blob/master/test/functional/transactions_tests.js#L35) for both of these cases?

Reply:
I have some ideas on how to do this, but I'd like to put it off until another change.

Reply:
that's fine, I'm actually refactoring that test runner for the next revision of txns changes